From 4e2ad9f42dec97fca409d593e8365cd1c59ddfee Mon Sep 17 00:00:00 2001
From: David Holsgrove <david.holsgrove@petalogix.com>
Date: Mon, 28 May 2012 22:46:54 +1000
Subject: [PATCH 59/79] microblaze: swapb / swaph insns for mcpu=8.30a require pattern compare

swapb and swaph instructions are introduced in v8.30a, but have an undocumented
dependence on -mxl-pattern-compare being set. This is due to be fixed in next
version, so the following checks are in place

mcpu < 8.30a; no swap insns

mcpu == 8.30a and -mxl-pattern-compare specified; swap insns allowed

mcpu > 8.30a; swap insns allowed

Signed-off-by: David Holsgrove <david.holsgrove@petalogix.com>
---
 gcc/config/microblaze/microblaze.c |   17 +++++++++++++----
 gcc/config/microblaze/microblaze.h |    2 +-
 2 files changed, 14 insertions(+), 5 deletions(-)

diff --git a/gcc/config/microblaze/microblaze.c b/gcc/config/microblaze/microblaze.c
index e3b30e9..2787c30 100755
--- a/gcc/config/microblaze/microblaze.c
+++ b/gcc/config/microblaze/microblaze.c
@@ -1828,20 +1828,29 @@ microblaze_option_override (void)
         microblaze_has_clz = 0;
     }
   ver = MICROBLAZE_VERSION_COMPARE (microblaze_select_cpu, "v8.30.a");
-  microblaze_has_swap = 1;
   if (ver < 0)
     {
         /* MicroBlaze prior to 8.30a didn't have swapb or swaph insns. */
-        microblaze_has_swap = 0;
         if (TARGET_REORDER)
           warning (0,
                  "-mxl-reorder can be used only with -mcpu=v8.30.a or greater");
     }
+  else if (ver == 0)
+   {
+        /* MicroBlaze v8.30a has an undocumented dependency on pattern compare for swapb / swaph insns. */
+        if (TARGET_PATTERN_COMPARE)
+        {
+            microblaze_has_swap = 1;
+            target_flags |= MASK_REORDER;
+        }
+    }
   else
    {
-       /* Add to Default target_flags if v8.30a or greater  */
-       target_flags |= MASK_REORDER;
+        /* Microblaze versions greater than v8.30a will be able to use swapb / swaph without pattern compare */
+        microblaze_has_swap = 1;
+        target_flags |= MASK_REORDER;
    }
+
   if (TARGET_MULTIPLY_HIGH && TARGET_SOFT_MUL)
     error ("-mxl-multiply-high requires -mno-xl-soft-mul");
 
diff --git a/gcc/config/microblaze/microblaze.h b/gcc/config/microblaze/microblaze.h
index 7868852..6bdea06 100755
--- a/gcc/config/microblaze/microblaze.h
+++ b/gcc/config/microblaze/microblaze.h
@@ -64,7 +64,7 @@ extern enum pipeline_type microblaze_pipe;
 #define TARGET_HAS_CLZ      (TARGET_PATTERN_COMPARE && microblaze_has_clz)
 
 /* Do we have SWAPB and SWAPH?  */
-#define TARGET_HAS_SWAP     (TARGET_REORDER && microblaze_has_swap)
+#define TARGET_HAS_SWAP     (microblaze_has_swap)
 
 /* The default is to not need GOT for TLS.  */
 #define TLS_NEEDS_GOT 0
-- 
1.7.1

